File: leet_code_algorithms/valid_palindrome/valid_palindrome.ts
// My Solution

var isPalindrome = (s:string) => {
  const lowCaseString = s.toLowerCase();
  let formattedS = "";
  let reversedS = "";
  for (let i = 0; i < lowCaseString.length; i++) {
    if (lowCaseString[i].match(/[0123456789a-z]/)) {
      formattedS += lowCaseString[i];
    }
  }
  for (let x = formattedS.length - 1; x > -1; x--) {
    reversedS += formattedS[x];
  }
  return formattedS === reversedS;
};

// ChatGPT Solution
var isPalindrome = function (s:string) {
  s = s.replace(/[\W_]/g, "").toLowerCase();

  let left = 0;
  let right = s.length - 1;

  while (right >= left) {
    if (s[left] !== s[right]) {
      return false;
    }

    right--;
    left++;
  }

  return true;
};

File: leet_code_algorithms/valid_palindrome/valid_palindrome.py
# My Solution
def isPalindrome(self, s: str) -> bool:
    validChars = {"a","b","c","d","e","f","g","h","i","j","k","l","m","n","o","p","q","r","s","t","u","v","w","x","y","z","0","1","2","3","4","5","6","7","8","9"}
    formattedString = ""
    for letter in s:
        if letter.lower() in validChars:
            formattedString += letter.lower()
    return formattedString == formattedString[::-1]

# ChatGPT Solution
def isPalindrome(self, s: str) -> bool:
    s = s.lower()
    valid = []
    for i in s:
        if i.isalpha() or i.isdigit():
            valid.append(i)
    return valid == valid[::-1]
